src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,387 @@
+            Node n = edge.node;
+            if (!dominates(newNode.getBlock().ptr, n.getBlock().ptr)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private final class CSEVisitor extends NodeVisitor.Default {
+
+        private final Map<TargetValue, Node> constCache = new HashMap<>();
+        private final Map<Entity, Node> addressCache = new HashMap<>();
+        private final Map<NodePreds, Node> nodeCache = new HashMap<>();
+
+        @RequiredArgsConstructor
+        private class NodePreds {
+            private final Node[] preds;
+            private final binding_irnode.ir_opcode opcode;
+            private final Mode mode;
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                NodePreds nodePreds = (NodePreds) o;
+                return Arrays.equals(preds, nodePreds.preds) && opcode == nodePreds.opcode && mode.equals(nodePreds.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                int result = Objects.hash(opcode, mode);
+                result = 31 * result + Arrays.hashCode(preds);
+                return result;
+            }
+        }
+
+        @Override
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    // replace and update cache
+                    nodeValues.put(n, node);
+                    nodeCache.put(np, node);
+                }
+            } else {
+                nodeCache.put(np, node);
+            }
+        }
+
+        private Node[] iterableToArray(Iterable<Node> iterable) {
+            List<Node> list = new ArrayList<>();
+            for(var entry : iterable) {
+                list.add(entry);
+            }
+            return list.toArray(new Node[0]);
+        }
+
+        @Override
+        public void visit(Add node) {
+            visitPreds(node, iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Add);
+        }
+
+        @Override
+        public void visit(And node) {
+            visitPreds(node, iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_And);
+        }
+
+        @Override
+        public void visit(Cmp node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Cond node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Proj proj) {
+            // do nothing
+        }
```

Comment:
I think you need to handle `Proj`s as well. For example, if you eliminate a pair of `Load`s, you'd also want to unify their result projections. Obviously beware that `Proj`s are only equal if they extract the same number.